segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,308 @@
+            throw new CompletionException(e);
+        }
+    }
+
+    private CompletableFuture<EventsReadAndTruncationPoints> readEventsForEventProcessor(EventProcessorImpl eventProcessor) {
+        return getLastProcessedEventOffset(eventProcessor.segment)
+                .thenApplyAsync(currentOffset -> readEventsFromSegment(eventProcessor, currentOffset), this.executor);
+    }
+
+    private EventsReadAndTruncationPoints readEventsFromSegment(EventProcessorImpl eventProcessor, int currentOffset) {
+        List<BufferView> readEvents = new ArrayList<>();
+        int finalTruncationOffset = currentOffset;
+        while (readEvents.size() < eventProcessor.getConfig().getMaxItemsAtOnce()) {
+            try {
+                int appendLength = bufferViewToInt(readAppendOfLength(eventProcessor.segment, finalTruncationOffset, Integer.BYTES));
+                finalTruncationOffset += Integer.BYTES;
+                BufferView data = readAppendOfLength(eventProcessor.segment, finalTruncationOffset, appendLength);
+                readEvents.add(data);
+                finalTruncationOffset += data.getLength();
+            } catch (Exception e) {
+                // We got some data, do not throw, just stop looping.
+                if (!readEvents.isEmpty()) {
+                    break;
+                }
+                // TimeoutException is expected if there is no data in the Segment. Otherwise it may be a problem.
+                throw (e instanceof TimeoutException) ? new NoDataAvailableException() : new CompletionException(e);
+            }
+        }
+        return new EventsReadAndTruncationPoints(readEvents, currentOffset, finalTruncationOffset);
+    }
+
+    private CompletableFuture<Void> updateAndTruncateInternalSegment(EventProcessorImpl ep,
+                                                                     EventsReadAndTruncationPoints readResult,
+                                                                     Timer iterationTime) {
+        return updateLastProcessedEventOffset(ep.segment, readResult.getFinalTruncationOffset())
+                .thenCompose(v -> ep.segment.truncate(readResult.getFinalTruncationOffset(), OPERATION_TIMEOUT))
+                .thenAccept(v -> {
+                    // Decrement outstanding bytes and publish the last latency value after successful truncation.
+                    ep.outstandingBytes.addAndGet(-readResult.getReadBytes());
+                    ep.lastIterationLatency.set(iterationTime.getElapsedMillis());
+                });
+    }
+
+    private BufferView readAppendOfLength(DirectSegmentAccess segment, int offset, int length) throws Exception {
+        @Cleanup
+        ReadResult readResult = segment.read(offset, length, OPERATION_TIMEOUT);
+        ReadResultEntry entry = readResult.next();
+        entry.requestContent(OPERATION_TIMEOUT);
+        return entry.getContent().get(OPERATION_TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
```

Comment:
1. do not use `.get` - same reason as why not use `.join`. Make this return a `CompletableFuture`.
2. Do not assume your entire read result will come from a single Read Result Entry. Depending on how much data is cached and what it reads from Storage, this may be multiple entries. You need an `AsyncReadResultProcessor` to properly process this `ReadResult`. (Tip: there may be some convenience methods on the `ReadResult` interface (be careful not to use any of the `VisibleForTesting` ones) that may help).

Reply:
You are right, sorry. Used `AsyncReadResultProcessor` to make the read async and chain the result of it to the rest of computation.